Signal/src/util/OWSBackup.m

```diff
@@ -0,0 +1,908 @@
+                             srcDirName:OWSBackup_AppSharedDataDirName]) {
+        return;
+    }
+
+    // TODO: Possibly verify database file location?
+
+    [OWSStorage storeDatabasePassword:[[NSString alloc] initWithData:databasePassword encoding:NSUTF8StringEncoding]];
+}
+
+- (nullable NSData *)readDataFromFileName:(NSString *)fileName
+{
+    OWSAssert(fileName.length > 0);
+    OWSAssert(self.backupDirPath.length > 0);
+    OWSAssert(self.encryptionKey);
+
+    NSString *filePath = [self.backupDirPath stringByAppendingPathComponent:fileName];
+
+    DDLogInfo(@"%@ readDataFromFileName: %@", self.logTag, filePath);
+
+    NSData *_Nullable encryptedData = [NSData dataWithContentsOfFile:filePath];
+    if (!encryptedData) {
+        OWSFail(@"%@ failed to read encrypted data: %@", self.logTag, fileName);
+        return nil;
+    }
+
+    NSData *_Nullable data = [Cryptography decryptAESGCMWithData:encryptedData key:self.encryptionKey];
+    if (!data) {
+        OWSFail(@"%@ failed to decrypt data: %@", self.logTag, fileName);
+        return nil;
+    }
+
+    return data;
+}
+
+- (BOOL)loadUserDefaults:(NSUserDefaults *)userDefaults fileName:(NSString *)fileName
+{
+    OWSAssert(userDefaults);
+    OWSAssert(fileName.length > 0);
+    OWSAssert(self.backupDirPath.length > 0);
+    OWSAssert(self.encryptionKey);
+
+    DDLogInfo(@"%@ loadUserDefaults: %@", self.logTag, fileName);
+
+    NSData *_Nullable data = [self readDataFromFileName:fileName];
+    if (!data) {
+        OWSFail(@"%@ Could not retrieve user defaults: %@.", self.logTag, fileName);
+        return NO;
+    }
+
+    NSError *error;
+    NSDictionary<NSString *, id> *_Nullable dictionary =
+        [NSKeyedUnarchiver unarchiveTopLevelObjectWithData:data error:&error];
+    if (!dictionary || error) {
+        OWSFail(@"%@ Could not unarchive user defaults: %@", self.logTag, error);
+        return NO;
+    }
+    if (![dictionary isKindOfClass:[NSDictionary class]]) {
+        OWSFail(@"%@ Unexpected archived user defaults: %@", self.logTag, error);
+        return NO;
+    }
+
+    // TODO: this doesn't yet remove any keys, so you end up with the "union".
+    for (NSString *key in dictionary) {
+        id value = dictionary[key];
+        OWSAssert(value);
+        [userDefaults setObject:value forKey:key];
+    }
+
+    return YES;
+}
+
+- (BOOL)restoreDirectoryContents:(NSString *)dstDirPath srcDirName:(NSString *)srcDirName
+{
+    OWSAssert(srcDirName.length > 0);
+    OWSAssert(dstDirPath.length > 0);
+    OWSAssert(self.backupDirPath.length > 0);
+
+    NSString *srcDirPath = [self.backupDirPath stringByAppendingPathComponent:srcDirName];
+
+    DDLogInfo(@"%@ restoreDirectoryContents: %@ -> %@", self.logTag, srcDirPath, dstDirPath);
+
+    if (![[NSFileManager defaultManager] fileExistsAtPath:srcDirPath]) {
+        DDLogInfo(@"%@ Skipping restore directory: %@.", self.logTag, srcDirPath);
```

Comment:
This seems like it should OWSFail -  or is there a normal reason that our backup would be missing a directory?

Reply:
Right, I'll add a comment:

```
Not all backups will have both a "app documents" and "shared data container" folder.  The latter should always be present for "modern" installs, but my intention is to be permissive here about what we accept so that we could easily apply this branch to historic (pre-shared data container) versions of the app and restore from them.
```